changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,552 @@
+
+Two HKeyrings WILL NOT have collisions if we append
+a Key Store ID parameter to the identifier
+(per issue
+[1](#issue-1-how-to-update-the-cache-identifier-for-the-hierarchy-keyring-to-allow-a-shared-cache)).
+
+Two CachingCMMs already allow shared cache.
+
+Per issue
+[4](#issue-4-should-we-add-a-prefix-to-all-cache-identifiers-in-hierarchy-keyrings-and-cachingcmms),
+a Hierarchy Keyring and a CachingCMM idenitifier will
+be distinct if we add the UTF8 encodings of prefixes
+'hierarchy_keyring' (17 bytes) and 'caching_cmm' (11 bytes)
+to each of them respectively.
+
+### Distinct sizes of the identifiers
+
+As an additional safety measure, we look closely
+at the formulae for cache identifiers in a Hierarchy Keyring
+and a CachingCMM encrypt and decrypt (note that these will
+actually be a total of 5, since the
+[cachingCMM encrypt cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/caching-cmm.md#appendix-a-cache-entry-identifier-formulas)
+has two different formulae based on if we
+specify the algorithm suite or not).
+
+In the
+[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula)
+Encrypt, as discussed before, we will remove the len_branch_key.
+For Decrypt, we will remove the len_branch_key
+and use the branch-key-digest in place of the branch-key-id.
+The size of the inputs to the external hash for both these
+cases then becomes:
+
+- Encrypt: 17 (prefix) + 64 + 1 + 6 = 88 bytes
+- Decrypt: 17 + 64 + 1 + 36 = 118 bytes
+
+Similarly, after making the proposed updates in the
+[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula),
+the size will be:
+
+- Encrypt without Algorithm Suite specified: 11 (prefix) + 64 + 1 + 64 = 140 bytes
+- Encrypt with Algorithm Suite specified: 11 + 64 + 1 + 2 + 64 = 142 bytes
+- Decrypt: 11 + 64 + 2 + 64\*x + 64 + 64 = 205 + 64x bytes ; where x is the number of EDKs in the decrypt string
+
+In addition to the explanations in the previous
+[subsection](#derivation-from-assumptions),
+all these input strings are also unique
+because they have different lengths. We just need
+to choose a good hashing algorithm which gives us a
+very large output space to project these strings to
+minimize collisions, and SHA512 will do the job.
+
+## Cache Identifier Formulae
+
+### Hierarchy Keyring Cache Identifier formula
+
+- [Encrypt](https://github.com/aws/aws-cryptographic-material-providers-library/blob/c07a51fc29ff70411f7573bca96d2a091db8c1ed/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/Keyrings/AwsKms/AwsKmsHierarchicalKeyring.dfy#L384) - SHA512(len_branch_key + SHA512(branch_key_id_utf8) + [0x00] + activeUtf8)[0..32] [[spec]](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/aws-kms/aws-kms-hierarchical-keyring.md#encryption-materials)
```

Comment:
len_branch_key is misleading. I believe it should refer to the length of the string branch-key-id, which is Uint32, as bytes

Reply:
That's a good point. However, we're going to remove the length of the branch key ID from the cache identifier because it doesn't add anything.